src/Interpreters/FillingRow.h

```diff
@@ -15,16 +15,20 @@ bool equals(const Field & lhs, const Field & rhs);
  */
 class FillingRow
 {
+    std::optional<Field> doJump(const FillColumnDescription & descr, size_t column_ind);
+    std::optional<Field> doLongJump(const FillColumnDescription & descr, size_t column_ind, const Field & to);
+
 public:
     explicit FillingRow(const SortDescription & sort_description);
 
     /// Generates next row according to fill 'from', 'to' and 'step' values.
     /// Return pair of boolean
     /// apply - true if filling values should be inserted into result set
     /// value_changed - true if filling row value was changed
-    std::pair<bool, bool> next(const FillingRow & to_row);
+    std::pair<bool, bool> next(const FillingRow & to_row, bool long_jump);
```

Comment:
Consider adding comment about new argument

Reply:
I changed the algorithm a bit because I noticed some problem in the output results.

There is now a shift method that tries to jump from the filling row to the next row from the data stream. To do this, it finds the first element in the sort order that is not equal to the next element of the original row, and makes a long jump to it (because there can be a large gap between values).

The difficulty is to figure out whether we need to generate some rows just before the next source row or not. Please check the `staleness with another regular with fill` test to get more clarity.